base/hashing2.jl

```diff
@@ -167,12 +167,35 @@ hash(x::Float16, h::UInt) = hash(float64(x), h)
 
 ## hashing collections ##
 const hashaa_seed = UInt === UInt64 ? 0x7f53e68ceb575e76 : 0xeb575e76
+const hashrle_seed = UInt == UInt64 ? 0x2aab8909bfea414c : 0xbfea414c
 function hash(a::AbstractArray, h::UInt)
     h += hashaa_seed
     h += hash(size(a))
-    for x in a
-        h = hash(x, h)
+
+    state = start(a)
+    done(a, state) && return h
+    x2, state = next(a, state)
+    done(a, state) && return hash(x2, h)
+
+    x1 = x2
+    while !done(a, state)
+        x1 = x2
+        x2, state = next(a, state)
+        if x2 == x1
+            # For repeated elements, use run length encoding
```

Comment:
I think we need to suck it up and do this correctly. It would be good to optimize `isequal` for Float64 though. Currently we have this:

```jl
julia> @code_llvm isequal(1.0, 1.0)

define i1 @julia_isequal_131700(double, double) {
top:
  %2 = bitcast double %0 to i64, !dbg !676
  %3 = bitcast double %1 to i64, !dbg !676
  %4 = fcmp uno double %0, 0.000000e+00, !dbg !676
  %5 = fcmp uno double %1, 0.000000e+00, !dbg !676
  %6 = and i1 %4, %5, !dbg !676
  %7 = icmp eq i64 %2, %3, !dbg !676
  %8 = or i1 %6, %7, !dbg !676
  ret i1 %8, !dbg !676
}
```

I wonder if there isn't something more efficient we could do. The challenges are:

- all `NaNs` are equal to each other
- `0.0` and `-0.0` are not equal to each other
- all other values are equal iff and only if they are bit-identical

Thanks, Obama.